### DISCUSSION TITLE
Asynchronous execution in sqlalchemy to run a query in non-blocking mode

### DISCUSSION THREAD

**Discussion by agl29** [sqlalchemy/sqlalchemy#6034] [Usage Questions]
**Is your feature request related to a problem? Please describe.**
At present the executeStatement call in sqlalchemy blocks until the query run is complete. As a result, no way to get progress information about the query. A non-blocking call will be very useful in this context, as it would return an OperationHandle immediately, which the client can use to track query progress. Query execution will proceed in a background thread.

**Describe the solution you'd like**
Please take a look at this https://issues.apache.org/jira/browse/HIVE-4617 Jira. If this is already in sqlalchemy can you please suggest to me a direction because I think with asyncio (`result = await conn.execute(select(t1).where(t1.c.name == "some name 1"))`) also we are waiting i.e. not getting query info till the completion?

**Have a nice day!**

**Comment by zzzeek** [comment]
hi there -

it sounds like a few things are going on here, one is the notion of getting "progress information" about a query, another is then the notion of running a query in a "background thread", and finally it looks like you're experimenting with asyncio and also comparing to the Apache Hive tool which is not so much a database library as a data warehouse over hadoop kind of thing, that is, it's written with a very specific notion of concurrency and program flow in mind.

So first off there is no mechanism I'm aware of in the Python DBAPI as well as the handful of async drivers available to get the "progress" of a SQL query, and in fact this is not something that standard relational databases generally support in their client APIs.  I can see how a platform like Hive built on MapReduce could do this, because that's the nature of a mapreduce job which is that you can track the individual bits of data through that process; it's largely transparent.   But if you are writing C code using [libpq](https://www.postgresql.org/docs/9.5/libpq.html) , to run a query to the database you're using a function like [PQexec](https://www.postgresql.org/docs/9.5/libpq-exec.html#LIBPQ-EXEC-MAIN) - it simply blocks. You can of course put it in a background thread, but that gives you no way to "check the progress" of the query, the background thread will be done when it's done and there's no way to get live introspection into that process.   This generally holds for all relational database client protocols.  Whether they are using blocking or non-blocking IO, there's only radio silence until the query begins to return rows, and when that part starts, you dont know how many rows there would be in any case.

next, being able to run blocking IO calls in background threads.  SQLAlchemy is not a concurrency framework, and there is no circumstance that SQLAlchemy itself would be spawning threads; that's out of its scope, and if you want to work with concurrency patterns such as these in Python using blocking IO database libraries, I would look into using the [concurrent.futures](https://docs.python.org/3/library/concurrent.futures.html) library which will give you that "operationhandle" concept you are looking for in the form of concurrent.futures.Executor.

Finally there's asyncio.   When using Python's asyncio system, now we use a different kind of database driver altogether, which is one that uses [non blocking sockets](https://docs.python.org/3/howto/sockets.html#non-blocking-sockets).   The non-blocking socket provides "non blocking" behavior in a different way than that of using a background thread.  With a background thread, the program has a main thread, then spins up the background thread, and in that background thread is where some operation proceeds, like a database operation that sends a command on a socket and waits for a response.   With non-blocking sockets, you emit the command on the non-blocking socket and it then returns immediately, not blocking anything.   The server that received the command, in this case the database, gets to take all the time it wants to produce a result, and the calling program need not spin up any thread at all.    in that sense it's already much more lightweight.  to actually get the answer back from the server it needs to [poll](https://docs.python.org/3/library/select.html) all of its non-blocking sockets to see which ones have some data available.

if you read the docs for asyncio, you won't see much discussion of those details as theyve been abstracted away underneath an event system that does the polling for us, and it then integrates into the control flow of the program such that our program has statements using the "await" keyword which means the program should stop running the next instruction and instead run other Python functions that may be proceeding concurrently.   if there's not anything else proceeding concurrently, then the program has nothing else to do but just wait for the one task to complete.

asyncio's usage requires that a few things get set up in order to see it work, including that the program is running in an [event loop](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio-event-loop) and that one or more [tasks](https://docs.python.org/3/library/asyncio-task.html) have been spun off.   When this is done, there is still no way to check the "progress" of any of the awaitables but certainly the program can do other things.

SQLAlchemy's [new asyncio system](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html) integrates into Python's asyncio system along with async DBAPIs, primarily the [asyncpg](https://magicstack.github.io/asyncpg/current/) DBAPI.  with this DBAPI, we have two options for running a query:

* execute the query in non blocking mode, and return the full result set in one step
* execute the query in non blocking mode using a server side cursor, then return when the cursor is ready with results.   the cursor can then be iterated in an asynchronous fashion as well.

So in the latter case, the querying process can be broken out into two phases, one where there's the initial wait for results and the second where result rows begin to come back.    On a database like PostgreSQL, if the query in question doesn't use ORDER BY , it can begin streaming result rows as they are located, so you can get some degree of "incremental query fetching", though for a lot of queries particularly if there's an ORDER BY, it has to line up most or all of the results first before yielding result rows out.

anyway, if you want to play with the non-blocking + server side cursor mode, SQLAlchemy's asyncio interface supports this using the [AsyncConnection.stream](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncConnection.stream) method for Core connections and [AsyncSession.stream](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncSession.stream) for ORM sessions.   When using stream, the AsyncResult object will work best when you have it fetch chunks of rows at a time with a method like [AsyncResult.partitions](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncResult.partitions).

the "threaded" APIs also support the notion of a server side cursor that you can "stream" from, general background on this is at https://docs.sqlalchemy.org/en/14/core/connections.html#engine-stream-results .

so to sum up:

1. relational database client libraries typically have a wait before results are returned
2. if the query uses ORDER BY, this will make it take longer for the first result rows to be yielded
3. both the blocking IO and non-blocking DBAPIs support the notion of a "server side cursor", where you can at least have control over how the result rows are fetched, once they become available.

**Comment by agl29** [comment]
Thanks for the reply @zzzeek 
cc @romainr

**Comment by romainr** [comment]
Thanks for all the detailed explanations @zzzeek !

And if I understood correctly, to summarize the "operationhandle" concept:

If does not exist natively as out of scope:
- cursor.poll() --> is dialect specific, in our case it was Hive/Presto: https://github.com/dropbox/PyHive#db-api-asynchronous
- Stream --> server side cursor when results are ready
- asyncio/SqlAlchemy async engine --> "under the cover" low level API for more efficient non blocking IO

Ideas for replicating a generic "operationhandle" with a DIY wrapper:
- https://docs.python.org/3/library/concurrent.futures.html on top of SqlAlchemy
- run the query into a separate Task (e.g. Celery)

**Reply by zzzeek** [reply]
cursor.poll(), yeah that's not a normal DBAPI thing  , and in fact SQLAlchemy isn't designed around any of the other cursor-level constructs like .scroll() either or even nextset().   if someone wants to work with these lesser used features of the cursor we advise them to work with the cursor directly.
